Adds input example and model signature to autologging for tf.keras for certain x types on tf.keras.Model.fit

bali0019 opened the pull request:
Signed-off-by: Jas Bali <bali0019@gmail.com>

## What changes are proposed in this pull request?

- Adds autologging input example/model signature support for certain x types: ndarray, dict mapping of input names to ndarrays, tf.data.Dataset and tf.keras.utils.Sequence. 
- Extracts either ndarray  or dic(str -> ndarray) for autologging input example/model signature
- Unit tests for all x types support added
## How is this patch tested?

Unit tests

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

## Release Notes

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Add example and model signature for autologging of tensorflow.keras flavor.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->
<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

bali0019 committed:
Add input example and model signature to autologging for tf.keras for certain x types on tf.keras.Model.fit

Signed-off-by: Jas Bali <bali0019@gmail.com>

bali0019 committed:
Temporary fix: Log warning for model_summary step that fails for tf.data.Dataset passed in as x for fit

Signed-off-by: Jas Bali <bali0019@gmail.com>

bali0019 committed:
Marked all tests as large type

Signed-off-by: Jas Bali <bali0019@gmail.com>

bali0019 commented:
@BenWilson2  
Questions/Concerns:
1) Do I need to add this for the new tests?
`@pytest.mark.skipif(
    Version(tf.__version__) < Version("2.6.0"),
    reason=("TensorFlow autologging is not used for vanilla Keras models in Keras < 2.6.0"),
)`

2) I added an exception catcher step inside ` mlflow/tensorflow/_autolog.py`  because `self.model.summary(print_fn=sum_list.append)` throws an exception for the new test model i added in unit tests. Need to investigate the cause of this further.

3) I disabled the log_model that happens inside keras callback (set it to False), registered inside `mlflow/tensorflow/__init__.py` since I added the `log_model` inside fit patch function that now logs input_example/model_signature as well for certain types, but I am not 100% sure if that is the right place to add the logic since we have that callback for `on_train_end`, where i didn't add the same logic, since even an input data slice can be big in certain cases. 

4) two new unit tests for x as tf.Data.Dataset are failing due to the shape (i believe), I am looking into this, but x as ndarray, dict and Sequence are good to go.

bali0019 committed:
Added fixes to unit tests for tf.data.Dataset x type

Signed-off-by: Jas Bali <bali0019@gmail.com>

bali0019 committed:
Reverted pylint comments in tests

Signed-off-by: Jas Bali <bali0019@gmail.com>

bali0019 committed:
Updated if/else with early return

Signed-off-by: Jas Bali <bali0019@gmail.com>

bali0019 committed:
Skipped new autologging tests for keras < 2.6.0

Signed-off-by: Jas Bali <bali0019@gmail.com>

BenWilson2 commented on mlflow/tensorflow/__init__.py:
```diff
@@ -50,14 +49,17 @@
 from mlflow.tracking._model_registry import DEFAULT_AWAIT_MAX_SLEEP_SECONDS
-
+from mlflow.models import infer_signature
+from mlflow.utils.autologging_utils import get_autologging_config
```
can add this to the block lines 49:57

bali0019 replied:
done. thanks!

BenWilson2 commented on mlflow/tensorflow/__init__.py:
```diff
@@ -887,14 +897,113 @@ def _patch_implementation(
 
                 history = original(inst, *args, **kwargs)
 
+                if log_models:
```
```suggestion
TF_SUPPORTED_AUTOLOGGING_TYPES = (np.ndarray, 
                  pd.Series, 
                  tensorflow.data.Dataset, 
                  dict, 
                  tensorflow.keras.utils.Sequence)
def _is_supported_for_autologging(input_data):
    support = isinstance(input_data, TF_SUPPORTED_AUTOLOGGING_TYPES)
    if support:
        if isinstance(input_data, dict):
            support = all({isinstance(value, np.ndarray) for value in input_data.values()})
        if isinstance(input_data, tensorflow.keras.utils.Sequence):
            support = isinstance(input_data[:][0], np.ndarray)
    if not support:
        warnings.warn("Tensorflow keras autologging only supports input types of: "
                      "numpy.ndarray, dict(<key> -> numpy.ndarray), tensorflow.data.Dataset, "
                      "or tensorflow.keras.utils.Sequence")
    return support
```
might want to simplify this logic a bit.

bali0019 replied:
Incorporated Corey's suggestion below and removed this method. thanks.

BenWilson2 commented on mlflow/tensorflow/__init__.py:
```diff
@@ -887,14 +897,113 @@ def _patch_implementation(
+                            )
+                        return is_x_supported
+
+                    def _get_input_data_slice(input_example, history):
```
does this shadow the higher level `history` variable? couldn't we just refer to it direclty with a history.params.get("steps", None) and then do a validate on the existence of that integer from within this private scoped function?

bali0019 replied:
good point, fixed. thanks.

BenWilson2 commented on mlflow/tensorflow/__init__.py:
```diff
@@ -887,14 +897,113 @@ def _patch_implementation(
+
+                    def _infer_model_signature(input_example):
+                        try:
+                            model_output = history.model.predict(input_example)
```
re: above comment, here you're referring to the encapsulated variable from the parent reference directly. Should be good to do above as well.

BenWilson2 commented:
> 3\. I disabled the log_model that happens inside keras callback (set it to False), registered inside `mlflow/tensorflow/__init__.py` since I added the `log_model` inside fit patch function that now logs input_example/model_signature as well for certain types, but I am not 100% sure if that is the right place to add the logic since we have that callback for `on_train_end`, where i didn't add the same logic, since even an input data slice can be big in certain cases.

I think that logic will have to be propagated (or moved) to ensure backwards compatibility with tf 2.6.3 and older versions (see cross version test results with "model not built yet" and other per-version specific issues where a validation of the instance __dict__() is looking for the values to be present. 
Seems to be just some minor things to address! Good job!

BenWilson2 commented:
autoformat

mlflow-automation committed:
Autoformat: https://github.com/mlflow/mlflow/actions/runs/1947424941

Signed-off-by: mlflow-automation <mlflow-automation@users.noreply.github.com>

dbczumar commented on mlflow/tensorflow/__init__.py:
```diff
@@ -887,14 +897,113 @@ def _patch_implementation(
+                if log_models:
+                    import numpy as np
+
+                    def _is_x_type_supported_for_autologging(input_data):
```
Do we need this function, or can we modify the behavior of `_get_input_data_slice` to return `None` if the input data is of an unsupported / unrecognized type for input example extraction?

dbczumar commented on mlflow/tensorflow/__init__.py:
```diff
@@ -887,14 +897,113 @@ def _patch_implementation(
+                    input_example = None
+                    if _is_x_type_supported_for_autologging(args[0]):
+                        input_example, signature = resolve_input_example_and_signature(
+                            lambda: _get_input_data_slice(args[0], history),
```
What happens if this throws? Should we try / catch and print a warning?

dbczumar replied:
Edit: Nevermind - `resolve_input_example_and_signature` handles this.

bali0019 committed:
Addressed PR reviews

Optimized imports
Simplified autologging logic for keras
Fixed cross-version failed unit tests

Signed-off-by: Jas Bali <bali0019@gmail.com>

bali0019 committed:
Merge branch 'autolog-input-example-signature-for-tf-keras' of github.com:bali0019/mlflow into autolog-input-example-signature-for-tf-keras

Signed-off-by: Jas Bali <bali0019@gmail.com>

bali0019 committed:
Fixed 2 pylint issues

Signed-off-by: Jas Bali <bali0019@gmail.com>

bali0019 committed:
Added a fix for failing cross-version tests for tf.keras in autologging

Signed-off-by: Jas Bali <bali0019@gmail.com>

bali0019 committed:
Added a fix for remainder of the failing cross-version tests for tf.keras in autologging

Signed-off-by: Jas Bali <bali0019@gmail.com>

BenWilson2 commented on mlflow/tensorflow/__init__.py:
```diff
@@ -887,14 +897,109 @@ def _patch_implementation(
+
+                    def _infer_model_signature(input_data_slice):
+                        try:
+                            original_stop_training = history.model.stop_training
```
no need to define this as a local variable. The conditional below could just use:
`if(Version(tensorflow.__version__) <= Version("2.1.4") and history.model.stop_training):`

BenWilson2 commented on mlflow/tensorflow/__init__.py:
```diff
@@ -887,14 +897,109 @@ def _patch_implementation(
+                            ):
+                                # For these versions, `stop_training` flag on Model is set to False
+                                # This flag is used by the callback
+                                # (inside `_log_early_stop_callback_metrics`)
```
nit: double backticks for method ref [970, 972, 975]

bali0019 replied:
done, thx.

BenWilson2 commented on tests/tensorflow/test_tensorflow2_autolog.py:
```diff
@@ -987,3 +1063,181 @@ def test_import_keras_with_fluent_autolog_enables_tensorflow_autologging():
+        for k, v in random_train_dict_mapping.items():
+            np.testing.assert_array_almost_equal(input_example[k], np.take(v, range(0, 5)))
+        pyfunc_model = mlflow.pyfunc.load_model(os.path.join(run.info.artifact_uri, "model"))
+        # Ensure input data slice works and is valid for the model
```
nit: remove note

bali0019 replied:
done, thx.

BenWilson2 commented on tests/tensorflow/test_tensorflow2_autolog.py:
```diff
@@ -49,6 +52,80 @@ def random_one_hot_labels():
+@pytest.fixture
+def keras_data_gen_sequence(random_train_data, random_one_hot_labels):
+    class DataGenerator(tf.keras.utils.Sequence):
+        # def __init__(self):
+        # self.batch_size = 32
+        # self.output_shape = (6, 12)
```
remove commented out lines

bali0019 replied:
done, thx.

BenWilson2 commented on tests/tensorflow/test_tensorflow2_autolog.py:
```diff
@@ -987,3 +1063,181 @@ def test_import_keras_with_fluent_autolog_enables_tensorflow_autologging():
+    input_example = _read_example(model_conf, model_path)
+    np.testing.assert_array_almost_equal(input_example, random_train_data[:5])
+    pyfunc_model = mlflow.pyfunc.load_model(os.path.join(run.info.artifact_uri, "model"))
+    # Ensure input data slice works and is valid for the model
```
nit: remove note

bali0019 replied:
done, thx.

BenWilson2 commented on tests/tensorflow/test_tensorflow2_autolog.py:
```diff
@@ -987,3 +1063,181 @@ def test_import_keras_with_fluent_autolog_enables_tensorflow_autologging():
+        model_conf = Model.load(os.path.join(model_path, "MLmodel"))
+        input_example = _read_example(model_conf, model_path)
+        pyfunc_model = mlflow.pyfunc.load_model(os.path.join(run.info.artifact_uri, "model"))
+        # Ensure input data slice works and is valid for the model
```
nit: remove note

bali0019 replied:
done, thx.

BenWilson2 reviewed (approved):
Great job! After a few nits are fixed LGTM!

bali0019 committed:
Cleaned up comments and commented out LOC

Signed-off-by: Jas Bali <bali0019@gmail.com>

dbczumar commented on mlflow/tensorflow/__init__.py:
```diff
@@ -887,14 +897,109 @@ def _patch_implementation(
                     callbacks = list(kwargs.get("callbacks") or [])
                     kwargs["callbacks"], self.log_dir = _setup_callbacks(
-                        callbacks, log_models, metrics_logger
+                        callbacks, False, metrics_logger
```
@bali0019 @BenWilson2 Does this imply that there's some cleanup we can do within `_setup_callbacks` to remove model logging code, since we're always passing `False` now? Can we file a follow-up PR for a cleanup here?

BenWilson2 replied:
for TF <2.1.0 the fit_generator() method patch still uses this flag. Should we mark this with a remove note when we no longer support those older TF versions?

dbczumar replied:
Does that imply that our autologging for signatures / input examples doesn't work with `fit_generator()`. If so, why not?

dbczumar commented on mlflow/tensorflow/__init__.py:
```diff
@@ -627,6 +629,8 @@ def autolog(
     disable_for_unsupported_versions=False,
     silent=False,
     registered_model_name=None,
+    log_input_examples=True,
```
@bali0019 @BenWilson2 This should default to `False` because input examples may contain sensitive data, and we want users to have to opt-in. Other autologging integrations default to `False` for this param.

dbczumar commented:
@bali0019 Thanks so much!
